sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/OpenAIClientBuilder.java

```diff
@@ -290,13 +306,63 @@ private HttpPipeline createHttpPipeline() {
+        return client;
+    }
+
+    private HttpPipeline createHttpPipelineNonAzureOpenAI() {
+        Configuration buildConfiguration =
+                (configuration == null) ? Configuration.getGlobalConfiguration() : configuration;
+        HttpLogOptions localHttpLogOptions = this.httpLogOptions == null ? new HttpLogOptions() : this.httpLogOptions;
+        ClientOptions localClientOptions = this.clientOptions == null ? new ClientOptions() : this.clientOptions;
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+        String clientName = PROPERTIES.getOrDefault(SDK_NAME, "UnknownName");
+        String clientVersion = PROPERTIES.getOrDefault(SDK_VERSION, "UnknownVersion");
+        String applicationId = CoreUtils.getApplicationId(localClientOptions, localHttpLogOptions);
+        policies.add(new UserAgentPolicy(applicationId, clientName, clientVersion, buildConfiguration));
+        policies.add(new RequestIdPolicy());
+        policies.add(new AddHeadersFromContextPolicy());
+        HttpHeaders headers = new HttpHeaders();
+        localClientOptions.getHeaders().forEach(header -> headers.set(header.getName(), header.getValue()));
+        if (headers.getSize() > 0) {
+            policies.add(new AddHeadersPolicy(headers));
+        }
+        this.pipelinePolicies.stream()
+                .filter(p -> p.getPipelinePosition() == HttpPipelinePosition.PER_CALL)
+                .forEach(p -> policies.add(p));
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+        policies.add(ClientBuilderUtil.validateAndGetRetryPolicy(retryPolicy, retryOptions, new RetryPolicy()));
+        policies.add(new AddDatePolicy());
+        policies.add(new CookiePolicy());
+        if (nonAzureOpenAIKeyCredential != null) {
+            policies.add(new NonAzureOpenAIKeyCredentialPolicy("Authorization", nonAzureOpenAIKeyCredential, "Bearer"));
+        }
+        this.pipelinePolicies.stream()
+                .filter(p -> p.getPipelinePosition() == HttpPipelinePosition.PER_RETRY)
+                .forEach(p -> policies.add(p));
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+        policies.add(new HttpLoggingPolicy(httpLogOptions));
+        HttpPipeline httpPipeline =
+                new HttpPipelineBuilder()
+                        .policies(policies.toArray(new HttpPipelinePolicy[0]))
+                        .httpClient(httpClient)
+                        .clientOptions(localClientOptions)
+                        .build();
+        return httpPipeline;
+    }
+
     /**
      * Builds an instance of OpenAIAsyncClient class.
      *
      * @return an instance of OpenAIAsyncClient.
      */
-    @Generated
     public OpenAIAsyncClient buildAsyncClient() {
+        if (nonAzureOpenAIKeyCredential != null) {
+            return new OpenAIAsyncClient(buildInnerNonAzureOpenAIClient());
+        }
```

Comment:
This could be me being overly paranoid or not understanding java concepts very well: I am a bit concerned about changes of internal state in this class that could come from different threads. Should we be annotating certain methods with `@Synchronized` ? I mean this, because I see a lot of null checks that could very easily become invalid should there be concurrent access to the object.

Reply:
good point. I can add it to the method
for double check, will wait for @srnagar

Reply:
after talking to Srikanta Builder is not guaranteed to be thread-safe. so will leave it as it is.